src/ucp/core/ucp_worker.c

```diff
@@ -2440,33 +2440,25 @@ ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self,
 static ucs_status_t
 ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self)
 {
-    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
-    ucp_request_t *req       = ucs_container_of(self, ucp_request_t, send.uct);
-    uct_ep_h uct_ep          = req->send.discard_uct_ep.uct_ep;
-    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
-    ucs_status_t status, status_add;
-
-    status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
-                          &req->send.state.uct_comp);
-    if (status == UCS_ERR_NO_RESOURCE) {
-        status_add = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
-        ucs_assert((status_add == UCS_ERR_BUSY) || (status_add == UCS_OK));
-        /* if added to the pending queue or to worker progress, need to remove
-         * the callback to not invoke it several times */
-        if (status_add == UCS_ERR_BUSY) {
-            uct_worker_progress_register_safe(worker->uct,
-                                              ucp_worker_discard_uct_ep_progress,
-                                              req, UCS_CALLBACKQ_FLAG_ONESHOT,
-                                              &cb_id);
-        }
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    uct_ep_h uct_ep    = req->send.discard_uct_ep.uct_ep;
+    ucs_status_t status;
 
-        return UCS_OK;
-    } else if (status == UCS_INPROGRESS) {
-        /* need to remove from the pending queue */
-        status = UCS_OK;
-    } else {
-        ucp_worker_discard_uct_ep_flush_comp(&req->send.state.uct_comp, status);
-    }
+    do {
+        status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
+                              &req->send.state.uct_comp);
+        if (status == UCS_ERR_NO_RESOURCE) {
+            status = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
+            ucs_assert((status == UCS_OK) || (status == UCS_ERR_BUSY));
+        } else if (status == UCS_INPROGRESS) {
+            /* need to remove from the pending queue */
+            status = UCS_OK;
+        } else {
+            ucs_assert(status != UCS_ERR_BUSY);
```

Comment:
this assert looks wrong since status value is returned by `flush` here

Reply:
this is not wrong, just to make sure that we will not enter the endless loop if `uct_ep_flush()` returns `UCS_ERR_BUSY`
added the comment

Reply:
but uct_ep_flush doc does not say that it cannot return `UCS_ERR_BUSY`

Reply:
> but uct_ep_flush doc does not say that it cannot return `UCS_ERR_BUSY`

yes, will handle it

@yosefe what do you think to use progress_register() instead of this loop?
seems to be error-prone or overcomplicated if it will have several `retrun`/`break`/`continue` in the loop